torch/csrc/jit/codegen/cuda/lower_warp_reduce.cpp

```diff
@@ -0,0 +1,546 @@
+        findLiveTvs(for_loop->body().exprs());
+        continue;
+      } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+        findLiveTvs(ite->thenBody().exprs());
+        findLiveTvs(ite->elseBody().exprs());
+        continue;
+      }
+
+      if (auto allocate = dynamic_cast<kir::Allocate*>(expr)) {
+        if (allocate->memoryType() == MemoryType::Local) {
+          if (auto kir_tv =
+                  dynamic_cast<kir::TensorView*>(allocate->buffer())) {
+            // We know only tvs that we'd want to consider are broadcast outputs
+            if (kir_tv->fuserTv()->definition()->isA<BroadcastOp>()) {
+              candidate_tvs_.push_back(kir_tv);
+              candidate_tv_set_.insert(kir_tv);
+            }
+          }
+        }
+      }
+
+      for (auto inp : expr->inputs()) {
+        if (auto ti = dynamic_cast<kir::TensorIndex*>(inp)) {
+          if (candidate_tv_set_.count(ti->view())) {
+            live_tvs_.insert(ti->view());
+          }
+        }
+      }
+    }
+  }
+
+  void findDeadTvs() {
+    for (auto tv : candidate_tvs_) {
+      if (!live_tvs_.count(tv)) {
+        dead_tvs_.insert(tv);
+      }
+    }
+  }
+
+  void eliminateDeadCode(const std::vector<kir::Expr*>& exprs) {
+    result_exprs_ = eliminateDeadCodeInScope(exprs);
+  }
+
+  bool shouldEliminate(kir::Expr* expr) {
+    if (auto allocate = dynamic_cast<kir::Allocate*>(expr)) {
+      if (auto buffer_tv = dynamic_cast<kir::TensorView*>(allocate->buffer())) {
+        if (dead_tvs_.count(buffer_tv)) {
+          return true;
+        }
+      }
+    } else if (auto broadcast = dynamic_cast<kir::BroadcastOp*>(expr)) {
+      if (auto out_ti = dynamic_cast<kir::TensorIndex*>(broadcast->out())) {
+        if (dead_tvs_.count(out_ti->view())) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  //! Returns a new vector of exprs with dead exprs
+  //!  eliminated.
+  std::vector<kir::Expr*> eliminateDeadCodeInScope(
+      const std::vector<kir::Expr*>& exprs) {
+    std::vector<kir::Expr*> result_exprs;
+
+    for (auto expr : exprs) {
+      auto result_expr = expr;
+      if (auto for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+        result_expr = eliminateDeadCode(for_loop);
+      } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+        result_expr = eliminateDeadCode(ite);
+      } else {
+        if (shouldEliminate(expr)) {
+          result_expr = nullptr;
+        }
+      }
+
+      // Push the result expr if not eliminated
+      if (result_expr) {
+        result_exprs.push_back(result_expr);
+      }
+    }
+
+    return result_exprs;
+  }
+
+  kir::ForLoop* eliminateDeadCode(kir::ForLoop* for_loop) {
+    auto new_loop_body = eliminateDeadCodeInScope(for_loop->body().exprs());
+    if (new_loop_body.empty()) {
+      return nullptr;
+    }
+
+    // TODO: we will need a kernel_ir cloner to make this
+    //  kind of logic re-usable.
```

Comment:
Agreed. Commented exactly the same in the other location.

Reply:
Discussed on thread:
https://github.com/csarofeen/pytorch/pull/976/files#r687139639